frame/bags-list/src/list/mod.rs

```diff
@@ -0,0 +1,670 @@
+		// easier; they can just configure `type BagThresholds = ()`.
+		let thresholds = T::BagThresholds::get();
+		let iter = thresholds.iter().copied();
+		let iter: Box<dyn Iterator<Item = u64>> = if thresholds.last() == Some(&VoteWeight::MAX) {
+			// in the event that they included it, we can just pass the iterator through unchanged.
+			Box::new(iter.rev())
+		} else {
+			// otherwise, insert it here.
+			Box::new(iter.chain(iter::once(VoteWeight::MAX)).rev())
+		};
+
+		iter.filter_map(Bag::get).flat_map(|bag| bag.iter())
+	}
+
+	/// Insert several ids into the appropriate bags in the list. Continues with insertions
+	/// if duplicates are detected.
+	///
+	/// Return the final count of number of ids inserted.
+	fn insert_many(
+		ids: impl IntoIterator<Item = T::AccountId>,
+		weight_of: impl Fn(&T::AccountId) -> VoteWeight,
+	) -> u32 {
+		let mut count = 0;
+		ids.into_iter().for_each(|v| {
+			let weight = weight_of(&v);
+			if Self::insert(v, weight).is_ok() {
+				count += 1;
+			}
+		});
+
+		count
+	}
+
+	/// Insert a new id into the appropriate bag in the list.
+	///
+	/// Returns an error if the list already contains `id`.
+	pub(crate) fn insert(id: T::AccountId, weight: VoteWeight) -> Result<(), Error> {
+		if Self::contains(&id) {
+			return Err(Error::Duplicate)
+		}
+
+		let bag_weight = notional_bag_for::<T>(weight);
+		let mut bag = Bag::<T>::get_or_make(bag_weight);
+		// unchecked insertion is okay; we just got the correct `notional_bag_for`.
+		bag.insert_unchecked(id.clone());
+
+		// new inserts are always the tail, so we must write the bag.
+		bag.put();
+
+		crate::CounterForListNodes::<T>::mutate(|prev_count| {
+			*prev_count = prev_count.saturating_add(1)
+		});
+
+		crate::log!(
+			debug,
+			"inserted {:?} with weight {} into bag {:?}, new count is {}",
+			id,
+			weight,
+			bag_weight,
+			crate::CounterForListNodes::<T>::get(),
+		);
+
+		Ok(())
+	}
+
+	/// Remove an id from the list.
+	pub(crate) fn remove(id: &T::AccountId) {
+		Self::remove_many(sp_std::iter::once(id));
+	}
+
+	/// Remove many ids from the list.
+	///
+	/// This is more efficient than repeated calls to `Self::remove`.
+	fn remove_many<'a>(ids: impl IntoIterator<Item = &'a T::AccountId>) {
+		let mut bags = BTreeMap::new();
+		let mut count = 0;
+
+		for id in ids.into_iter() {
+			let node = match Node::<T>::get(id) {
+				Some(node) => node,
+				None => continue,
+			};
+			count += 1;
+
+			if !node.is_terminal() {
+				// this node is not a head or a tail and thus the bag does not need to be updated
+				node.excise()
+			} else {
+				// this node is a head or tail, so the bag needs to be updated
+				let bag = bags
+					.entry(node.bag_upper)
+					.or_insert_with(|| Bag::<T>::get_or_make(node.bag_upper));
+				// node.bag_upper must be correct, therefore this bag will contain this node.
+				bag.remove_node_unchecked(&node);
+			}
+
+			// now get rid of the node itself
+			crate::ListNodes::<T>::remove(id);
```

Comment:
might be nicer to hide this behind a function in `struct Node`, to not touch an unrelated storage item directly. Ideally, only `Node` touches `ListNodes`.

Reply:
651148e6e2ee6e2bbf2679f220094b367278e449